libs/parallelism/executors/include/hpx/executors/sender_receiver_executor.hpp

```diff
@@ -0,0 +1,244 @@
+        // with this executor.
+        using executor_parameters_type = static_chunk_size;
+
+        template <typename T, typename... Ts>
+        using future_type = hpx::future<T>;
+
+        // NonBlockingOneWayExecutor interface
+        template <typename F, typename... Ts>
+        void post(F&& f, Ts&&... ts)
+        {
+            schedule(sched_) |
+                transform(detail::captured_args(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)) |
+                detach();
+        }
+
+        // OneWayExecutor interface
+        template <typename F, typename... Ts>
+        decltype(auto) sync_execute(F&& f, Ts&&... ts)
+        {
+            return schedule(sched_) |
+                transform(detail::captured_args(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)) |
+                sync_wait();
+        }
+
+        // TwoWayExecutor interface
+        template <typename F, typename... Ts>
+        decltype(auto) async_execute(F&& f, Ts&&... ts)
+        {
+            return schedule(sched_) |
+                transform(detail::captured_args(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)) |
+                make_future();
+        }
+
+        template <typename F, typename Future, typename... Ts>
+        decltype(auto) then_execute(F&& f, Future&& predecessor, Ts&&... ts)
+        {
+            return std::forward<Future>(predecessor) | keep_future() |
+                on(sched_) |
+                transform(detail::captured_args_unary(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)) |
+                make_future();
+        }
+
+        // BulkTwoWayExecutor interface
+        template <typename F, typename S, typename... Ts>
+        decltype(auto) bulk_async_execute(F&& f, S const& shape, Ts&&... ts)
+        {
+            using shape_element = typename std::iterator_traits<typename hpx::
+                    traits::range_traits<S>::iterator_type>::value_type;
+            using result_type =
+                hpx::util::invoke_result_t<F, shape_element, Ts...>;
+
+            std::vector<hpx::future<result_type>> results;
+            results.reserve(hpx::util::size(shape));
+
+            for (auto const& s : shape)
+            {
+                results.push_back(schedule(sched_) |
+                    transform(detail::captured_args(f, s, ts...)) |
+                    make_future());
+            }
+
+            return results;
+        }
+
+        template <typename F, typename S, typename... Ts>
+        decltype(auto) bulk_sync_execute(F&& f, S const& shape, Ts&&... ts)
+        {
+            return schedule(sched_) |
+                bulk(shape,
+                    detail::captured_args_unary(
+                        std::forward<F>(f), std::forward<Ts>(ts)...)) |
+                sync_wait();
+        }
+
+        // FIXME: implement bulk_then_execute
```

Comment:
This is supposed to return a future of a vector (container) of the results of each invocation of `f`, correct? If yes, would this do:

```
return just_on(sched_, std::vector<result_type>(size(shape))) |
  bulk(shape, [=](auto i, std::vector<result_type> v) {
    v[i] = HPX_INVOKE(f, ts...);
  }) | 
  make_future();
```
(ignoring forwarding, iterators vs indices etc.)?

If no, then maybe not... Could you expand on what you had in mind for what we could do differently/better than the current emulation?

Reply:
What about running things concurrently? Also, it should depend on the future passed in. Otherwise, yes - the function should return a `future<vector<result_type>>`, where `result_type` is whatever `F` returns (or simply a `future<void>` if `F` returns `void`)..

Reply:
> What about running things concurrently?

That should be taken care of by `bulk`, no? There are of course two routes for this, 1. use the emulation fallback which uses `async_execute` and 2. use `bulk`. Since this is `(P2300/P0443)scheduler_executor` I think it makes most sense to use `bulk` and if there's no customization of `bulk` for the scheduler and it does a serial for-loop then so be it.

> Also, it should depend on the future passed in.

Uh, of course... something like this:

```
return when_all(keep_future(predecessor), just(std::vector<result_type>(size(shape)))) |
  on(sched_) |
  bulk(shape, [=](auto i, auto predecessor, std::vector<result_type> v) {
    v[i] = HPX_INVOKE(f, predecessor.get(), ts...);
  }) |
  transform([](auto future, std::vector<result_type> v) {
    return v;
  }) |
  make_future();
```
?

That should:
- run when the predecessor future is ready
- pass the future value and the additional arguments to f
- return only the vector of results as a future

Reply:
> > What about running things concurrently?
> 
> That should be taken care of by `bulk`, no? There are of course two routes for this, 1. use the emulation fallback which uses `async_execute` and 2. use `bulk`. Since this is `(P2300/P0443)scheduler_executor` I think it makes most sense to use `bulk` and if there's no customization of `bulk` for the scheduler and it does a serial for-loop then so be it.

I assumed that `bulk()` would simply run things sequentially as it does not take a scheduler as one of its argument (at least now). Once we introduce the `get_completion_scheduler` CPO that could be used for this purpose, however.

Reply:
Allowing `bulk()` to run things concurrently would also assume that it is fork/join. Not sure if using it in this case would be a good idea....

Reply:
> I assumed that bulk() would simply run things sequentially as it does not take a scheduler as one of its arguments (at least now). Once we introduce the get_completion_scheduler CPO that could be used for this purpose, however.

Yes, the dispatching has to be changed to allow a scheduler overload to be used as well, but that's exactly the plan in my understanding. Running things sequentially is just meant to be the default.

> Allowing `bulk()` to run things concurrently would also assume that it is fork/join. Not sure if using it in this case would be a good idea....

What other option is there (ok, running things sequentially...)? Or why would it not be a good idea? If one doesn't want fork/join or concurrency one should pick a scheduler that doesn't allow concurrency (`inline_scheduler` or something like that?).